src/format_rules.ml

```diff
@@ -0,0 +1,101 @@
+open Import
+
+let flag_of_kind : Ml_kind.t -> _ =
+  function
+  | Impl -> "--impl"
+  | Intf -> "--intf"
+
+let config_includes (config : Dune_file.Auto_format.t) s =
+  match config.enabled_for with
+  | Default -> true
+  | Only set -> List.mem s ~set
+
+let add_diff sctx loc alias ~dir input output =
+  let module SC = Super_context in
+  let open Build.O in
+  let action = Action.diff input output in
+  SC.add_alias_action sctx alias ~loc:(Some loc) ~locks:[] ~stamp:input
+    (Build.paths [input; output]
+     >>>
+     Build.action
+       ~dir
+       ~targets:[]
+       action)
+
+let gen_rules sctx (config : Dune_file.Auto_format.t) ~dir =
+  let loc = config.loc in
+  let files =
+    File_tree.files_of
+      (Super_context.file_tree sctx)
+      (Path.drop_build_context_exn dir)
+  in
+  let subdir = ".formatted" in
+  let output_dir = Path.relative dir subdir in
+  let alias = Build_system.Alias.make "fmt" ~dir in
+  let alias_formatted = Build_system.Alias.make "fmt" ~dir:output_dir in
+  let resolve_program = Super_context.resolve_program sctx ~loc:(Some loc) in
+  let setup_formatting =
+    (fun file (arrows_acc, extra_deps_acc) ->
+       let input_basename = Path.basename file in
+       let input = Path.relative dir input_basename in
+       let output = Path.relative output_dir input_basename in
+
+       let ocaml kind =
+         if config_includes config Ocaml then
+           let exe = resolve_program "ocamlformat" in
+           let args =
+             let open Arg_spec in
+             [ A (flag_of_kind kind)
+             ; Dep input
+             ; A "--name"
+             ; Path file
```

Comment:
`Path.t` values in argument specification eventually get translated to: `Path.reach <path> ~from:<dir-where-the-command-is-executed>`. So for a file `src/foo.ml`, I believe this will get turned to `../../../src/foo.ml` given that the command is run in `_build/default/src`. I suggest to execute the command from `_build/default` (available via `(Super_context.context sctx).build_dir` and simply pass `input` here.

Reply:
Yes, that's what's passed (this appears in the test output). I'm not sure what we gain from changing the directory. This has to refer to both the build and the source directory, so we'll have to do some path manipulation anyway.